expo/expo
# feat(cli): add development code signing

# Why

This adds development code signing which will supersede legacy code signing. For now clients will continue to support and request both, but after a while we'll stop requesting the old style and then a while after that we'll stop supporting legacy code signing on this dev server.

The rough strategy for this is outlined in https://github.com/expo/code-signing-certificates/pull/1.

Support for development certificates was added to the server in https://github.com/expo/universe/pull/9232 and https://github.com/expo/universe/pull/9349.

Support for chains (required for this feature) was added to the client in https://github.com/expo/expo/pull/16375 and https://github.com/expo/expo/pull/16634.

Depends on https://github.com/expo/expo/pull/16804.

Closes ENG-4480.

# How

Inline docblocks and comments are probably the best way to describe the how, but the general approach is:
1. Request comes in with `keyid=expo-root` or `keyid=<user keyid>`, we handle each one separately.
2. We sign the manifest with the private key and send the signature back.

# Test Plan

Run new and existing tests.

Manual tests:
1. Load via curl, see correct response including cert chain and signature: `curl --request GET \
  --url 'http://192.168.0.113:19000/?platform=ios' \
  --header 'accept: multipart/mixed' \
  --header 'expo-expect-signature: sig, keyid="expo-root", alg="rsa-v1_5-sha256"'`
2. Open in expo go with debugger attached, see code signing works by breakpointing where it validates the signature: `nexpo start --force-manifest-type=expo-updates`
